### PR TITLE
Wip mds snaps

### PR DESCRIPTION
**Pull request by ukernel**

### CODE REVIEW

File: src/mds/StrayManager.cc [hunk 1]
```diff
@@ -637,23 +637,29 @@ void StrayManager::eval_remote_stray(CDentry *stray_dn, CDentry *remote_dn)
 {
   assert(stray_dn != NULL);
   assert(stray_dn->get_dir()->get_inode()->is_stray());
+  CDentry::linkage_t *stray_dnl = stray_dn->get_projected_linkage();
```
Comment (jcsp) [inline src/mds/StrayManager.cc]: Are stray_dnl/stray_in just moved to the outer scope in order to run the asserts in more cases?

File: src/mds/StrayManager.cc [hunk 2]
```diff
@@ -637,23 +637,29 @@ void StrayManager::eval_remote_stray(CDentry *stray_dn, CDentry *remote_dn)
...
-
     if (!stray_in->remote_parents.empty()) {
-      remote_dn = *stray_in->remote_parents.begin();
-    } else {
-      dout(20) << __func__
-        << ": not reintegrating (no remote parents in cache)" << dendl;
+      for (compact_set<CDentry*>::iterator p = stray_in->remote_parents.begin();
+	   p != stray_in->remote_parents.end();
+	   ++p)
+	if ((*p)->last == CEPH_NOSNAP) {
```
Comment (jcsp) [inline src/mds/StrayManager.cc]: It makes sense to be intuitively that we wouldn't want to reintegrate into a snapshotted remote dentry, but I'm curious about what the actual problem is with doing it?  Do we just not have the infrastructure for that kind of modification?

File: src/mds/StrayManager.cc (hunk 1 above)
Comment (ukernel) [inline src/mds/StrayManager.cc]: yes

File: src/mds/StrayManager.cc (hunk 2 above)
Comment (ukernel) [inline src/mds/StrayManager.cc]: The problem is that the rename request creates a null dentry with wrong 'first'. the null dentry will confuse later snap dentry lookup. (that bug is fixed by commit 'mds: properly set null dentry's first'
in this RP)

**Comment by gregsfortytwo** [comment]
I am distressed at not being able to rely on bloom for snap dentries, although I understand the problem. Please create some tickets about that — I'm thinking one for "bloom filters can't handle snap dentries" and one for "allow us to to keep using bloom filters on HEAD entries even if snapshots are evicted".

**Comment by gregsfortytwo** [comment]
@jcsp, please finish reviewing and merge or toss back to @ukernel as appropriate. :)

**Comment by gregsfortytwo** [comment]
poke @jcsp

**Pull request merged**